dialogue-core/src/main/java/com/palantir/dialogue/core/RetryingChannel.java

```diff
@@ -0,0 +1,88 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.SettableFuture;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+
+/**
+ * Retries calls to the underlying channel upon failure.
+ */
+public final class RetryingChannel implements Channel {
+    private static final Executor DIRECT_EXECUTOR = MoreExecutors.directExecutor();
+    private static final int DEFAULT_MAX_RETRIES = 4;
+
+    private final Channel delegate;
+    private final int maxRetries;
+
+    public RetryingChannel(Channel delegate) {
+        this(delegate, DEFAULT_MAX_RETRIES);
+    }
+
+    @VisibleForTesting
+    RetryingChannel(Channel delegate, int maxRetries) {
+        this.delegate = delegate;
+        this.maxRetries = maxRetries;
+    }
+
+    @Override
+    public ListenableFuture<Response> createCall(Endpoint endpoint, Request request) {
+        SettableFuture<Response> future = SettableFuture.create();
+
+        Supplier<ListenableFuture<Response>> callSupplier = () -> delegate.createCall(endpoint, request);
+        FutureCallback<Response> retryer = new RetryingCallback<>(callSupplier, future);
+        Futures.addCallback(callSupplier.get(), retryer, DIRECT_EXECUTOR);
+
+        return future;
+    }
+
+    private final class RetryingCallback<T> implements FutureCallback<T> {
+        private final AtomicInteger failures = new AtomicInteger(0);
+        private final Supplier<ListenableFuture<T>> runnable;
+        private final SettableFuture<T> delegate;
+
+        private RetryingCallback(Supplier<ListenableFuture<T>> runnable, SettableFuture<T> delegate) {
+            this.runnable = runnable;
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void onSuccess(T result) {
+            delegate.set(result);
+        }
+
+        @Override
+        public void onFailure(Throwable throwable) {
+            if (failures.incrementAndGet() < maxRetries) {
+                Futures.addCallback(runnable.get(), this, DIRECT_EXECUTOR);
```

Comment:
not all requests can be retried, probably need to worry about that here

Reply:
so I think the retryability of an endpoint should be specified in the `Endpoint` definition above, at which point we can either wire up this call back or not [here](https://github.com/palantir/dialogue/pull/75/files#diff-3b3ed2f3bd0b0b9d1adce9a9fa7ab891R59)

Reply:
actually, there are two reasons a request may not be retryable: first, if the endpoint statically delcares it (could be derived from the API def), or second when the request carries read-once data like streams. the latter cannot be derived from the endpoint alone.

Reply:
I think we're going to need to define whether or not the request body is streamed in the endpoint, or possibly as part of the `RequestBody`. We should also make it part of the contract of `RequestBody#content` that each time it's called it must hand back a new `InputStream`

Reply:
yes it could be part of the request body.